### ISSUE TITLE
Detect and track card pack openings

### ISSUE THREAD

**Issue by culinko** [ko-vasilev/Hearthstone-Collection-Tracker#24]
Hello once again :)

We've been talking with @Epix37 about logging pack openings and he suggested that this feature could be a viable part of your plugin. Recently, we've found out that when you open your pack, the info gets stored in the log files! It looks like this: http://hastebin.com/dojugodaci.hs. The only things you'd need to watch out for are obtaining cards through levelling up (normal, golden), beating adventure bosses, crafting cards, opening seasonal chests and obtaining cards directly as arena rewards (not sure if there are more). Because all of them are stored in Achievements.log, along with the cards from the collection that you already own (when you start hs, so no need for manually importing them from the game!). Also there could be a setting to "auto add" every new card you open into your collection.

What do you think, are you interested in this kind of feature? (I will be saving my Achievements.log with the pack openings manually for now, so I hope there will be a way to import the info into the collection tracker once this feature is implemented, that is if you decide to include it).

**Comment by ko-vasilev** [comment]
Hey thank you so much for sharing with me!
Yes it is a nice feature to implement, I'll look what can I do with it.
BTW for now if you'll be opening a bunch of TGT cards on release you can just import them from the game (you can select specific set to import, not entire collection).

**Comment by culinko** [comment]
Oh I didn't mean importing the cards into the collection, I meant importing the pack opening info. When I was talking about tracking pack openings, I wasn't just talking about tracking cards you don't have in the collection, I was thinking about tracking every pack opening somewhere, so you could get some useful stats later, like how many packs you opened, how many legendaries you opened, with all the percentages and stuff like that ;)

**Comment by ko-vasilev** [comment]
Oh that's what you meant. Yes this will be fun to analyze in the future and maybe make like TOP PACK PULL OF THE DAY :D
I still think that if I implement something like this, there should be an option of automatically adding opened cards into collection (or at least ask user about this?).

**Comment by culinko** [comment]
Yeah, you can also highlight the best pack you've ever opened (by dust or rarity). Store relative dust value of pack (only count the cards you already have in collection as dust) and also absolute dust value of pack (count all cards as dust). There are endless possibilities of what to do with the data once you obtain them.

Yes, definitely an option and also separate the options for normal and golden cards, because some people disenchant all golden cards right away and don't want them in the collection.

**Comment by culinko** [comment]
Alright, I'm getting an echo that Achievements.log in fact doesn't contain list of all the cards you own. Just non-craftable cards (basic and adventure) cards (no expert, GVG or TGT cards). But maybe some other logger contains them, I will check it in a while.

**Comment by ko-vasilev** [comment]
Well, that escalated quickly...
Who knows, maybe this will be possible in future :)

**Comment by culinko** [comment]
Yeah, too bad that you can't auto-import your collection from the logs. But luckily the pack opening thingy still stands :)

**Comment by djkaty** [comment]
It only contains the universal basic cards, the level 1-10 cards and cards unlocked from adventures.

My strategy for pack openings is to ignore anything that isn't preceeded by [Bob] Network.OpenBooster

**Comment by jleclanche** [comment]
A better solution is probably to rely on timestamps, and filter out any block which contains non-purchaseable cards.

**Comment by culinko** [comment]
Oh right, you can block any non-purchaseable cards, that would leave out only cards from arena rewards, crafting and seasonal rewards from chests. But you can't get 5 of those in a couple of milliseconds anyway.

**Comment by culinko** [comment]
So, this is an example of the classic pack I opened recently. The .log files have a slightly different format than output_log.txt, but it should give you a basic idea of the log structure. The contents are: NotifyOfCardGained, card name, card ID, card type (minion, ability, weapon), normal/golden and number of copies you have in collection (including the one you just opened).

D 22:10:04.2179163 NotifyOfCardGained: [name=Windfury Harpy cardId=EX1_033 type=MINION] [CardFlair: Premium=STANDARD] 5
D 22:10:04.2199164 NotifyOfCardGained: [name=Blade Flurry cardId=CS2_233 type=ABILITY] [CardFlair: Premium=STANDARD] 3
D 22:10:04.2229165 NotifyOfCardGained: [name=Betrayal cardId=EX1_126 type=ABILITY] [CardFlair: Premium=STANDARD] 6
D 22:10:04.2259165 NotifyOfCardGained: [name=Silence cardId=EX1_332 type=ABILITY] [CardFlair: Premium=STANDARD] 3
D 22:10:04.2279166 NotifyOfCardGained: [name=Faerie Dragon cardId=NEW1_023 type=MINION] [CardFlair: Premium=STANDARD] 4

So as you can see, I already had all of the cards. So the "relative" dust value was 40 dust and absolute value was also 40 dust.

Let's take a look at another pack:

D 21:52:45.4240903 NotifyOfCardGained: [name=Undercity Valiant cardId=AT_030 type=MINION] [CardFlair: Premium=STANDARD] 2
D 21:52:45.4240903 NotifyOfCardGained: [name=Sparring Partner cardId=AT_069 type=MINION] [CardFlair: Premium=STANDARD] 1
D 21:52:45.4240903 NotifyOfCardGained: [name=Buccaneer cardId=AT_029 type=MINION] [CardFlair: Premium=STANDARD] 3
D 21:52:45.4290905 NotifyOfCardGained: [name=Evil Heckler cardId=AT_114 type=MINION] [CardFlair: Premium=STANDARD] 3
D 21:52:45.4290905 NotifyOfCardGained: [name=Bash cardId=AT_064 type=ABILITY] [CardFlair: Premium=STANDARD] 2

As you can see, the "relative" value was 10 dust (because there were only 2 duplicate commons I had both copies of) and the absolute value was 40 dust (1 rare and 4 commons).

So, everytime you open a card and the number is higher than 2 (for legendaries higher than 1), count that card as a duplicate (or you can deduct that info already from the user's collection anyway). Only duplicate cards count to the relative dust count.

So, it would be awesome if you could list all the packs you've ever opened in some kind of UI, also with both relative and absolute dust counts and it would be automatically tracked by HDT every time you opened a pack :)

**Comment by ko-vasilev** [comment]
Yes, I'll look what can be done in this direction. Likely, I saved my log from GVG packs opening :)

**Comment by jleclanche** [comment]
If you need more card opening logs, you can ping me. We gathered hundreds for the blog post: http://hearthsim.info/blog/the-grand-tournament-card-pack-opening/

**Comment by ko-vasilev** [comment]
@jleclanche Thank you!

**Comment by culinko** [comment]
Blizzard _might_ have added back some of the logging, because today I crafted 3 cards and this was shown in Achievements.log:

```
D 17:47:36.1533550 NotifyOfCardGained: [name=Ragnaros, Lightlord cardId=OG_229 type=MINION] NORMAL 1
D 17:47:43.6936355 NotifyOfCardGained: [name=Nat Pagle cardId=EX1_557 type=MINION] NORMAL 1
D 17:47:49.7468607 NotifyOfCardGained: [name=Herald Volazj cardId=OG_316 type=MINION] NORMAL 1
```
